go/gpg/main.go

```diff
@@ -0,0 +1,138 @@
+	secretName = flag.String("secret", "GPG_PRIVATE_KEY", "GitHub organization secret name")
+	visibility = flag.String("visibility", "all", "GitHub organization secret visibility")
+
+	gpgName    = flag.String("gpg-name", "opentofu", "GPG name")
+	gpgComment = flag.String("gpg-comment", "This is the key used to sign opentofu providers", "GPG comment")
+	gpgEmail   = flag.String("gpg-email", "your.email@example.com", "GPG comment")
+
+	pat string
+)
+
+func main() {
+	pat = os.Getenv("GITHUB_PAT")
+	if pat == "" {
+		panic("GITHUB_PAT environment variable not set")
+	}
+
+	flag.Parse()
+
+	config := &packet.Config{
+		DefaultCipher:          packet.CipherAES256,
+		DefaultCompressionAlgo: packet.CompressionZLIB,
+		CompressionConfig: &packet.CompressionConfig{
+			Level: packet.BestCompression,
+		},
+		RSABits: 4096,
+	}
+
+	es, err := openpgp.NewEntity(*gpgName, *gpgComment, *gpgEmail, config)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	var keyLifetime uint32 = 3 * 365 * 24 * 60 * 60
+
+	for _, id := range es.Identities {
+		id.SelfSignature = &packet.Signature{
+			CreationTime:              time.Now(),
+			KeyLifetimeSecs:           &keyLifetime,
+			SigType:                   packet.SigTypePositiveCert,
+			PubKeyAlgo:                packet.PubKeyAlgoRSA,
+			Hash:                      config.Hash(),
+			IssuerKeyId:               &es.PrimaryKey.KeyId,
+			PreferredHash:             []uint8{8},
+			PreferredCompression:      []uint8{2},
+			PreferredSymmetric:        []uint8{9},
+			IsPrimaryId:               new(bool),
+			FlagsValid:                true,
+			FlagCertify:               false,
+			FlagEncryptCommunications: false,
+			FlagEncryptStorage:        false,
+		}
+
+		err := id.SelfSignature.SignUserId(id.UserId.Id, es.PrimaryKey, es.PrivateKey, config)
+		if err != nil {
+			log.Fatalf("Error signing identity: %v", err)
+		}
+	}
+
+	privateKeyBuffer := new(bytes.Buffer)
+	err = es.SerializePrivate(privateKeyBuffer, config)
+	if err != nil {
+		log.Fatalf("Error serializing key: %v", err)
+	}
+
+	err = setSecret(context.Background(), privateKeyBuffer.Bytes())
+	if err != nil {
+		log.Fatalf("Error uploading GPG key to GitHub: %v", err)
+	}
+}
+
+func setSecret(ctx context.Context, privKey []byte) error {
+	log.Println("Setting secret in", *org, "...")
+
+	g := github.NewTokenClient(nil, pat)
+	k, _, err := g.Actions.GetOrgPublicKey(ctx, *org)
+	if err != nil {
+		return fmt.Errorf("could not get key: %w", err)
+	}
+
+	en, err := encodeWithPublicKey(base64.StdEncoding.EncodeToString(privKey), *k.Key)
+	if err != nil {
+		return fmt.Errorf("could not encode: %w", err)
+	}
+
+	_, err = g.Actions.CreateOrUpdateOrgSecret(ctx, *org, &github.EncryptedSecret{
+		Name:           *secretName,
+		KeyID:          *k.KeyID,
+		EncryptedValue: en,
+		Visibility:     *visibility,
+	})
+
+	return err
+}
+
+func encodeWithPublicKey(text string, publicKey string) (string, error) {
```

Comment:
I don't understand. Why are we encrypting the newly generated private gpg key with some random public key? Is this some GitHub Secrets API-specific thing?

Reply:
Anyhow, could use a comment

Reply:
>Is this some GitHub Secrets API-specific thing?

You're totally right here. This is github specific. It's basically how you store a secret.